### PR TITLE
Don't include any TLSv1.3 ciphersuites that are disabled

### PR DESCRIPTION
**Pull request by olszomal**
The CHACHA20_POLY1305 cipher is disabled in the FIPS mode, but s_server is able to send TLS_CHACHA20_POLY1305_SHA256 cipher suite in the ServerHello message:
```
$ openssl s_server -accept 127.0.0.1:4433 -cert cert.pem -provider fips -propquery fips=yes -ciphersuites TLS_CHACHA20_POLY1305_SHA256
Using default temp DH parameters
ACCEPT
ERROR
shutting down SSL
CONNECTION CLOSED
```
```
$ openssl s_client -connect 127.0.0.1:4433
Connecting to 127.0.0.1
CONNECTED(00000003)
40970115F87F0000:error:0A0001BB:SSL routines:ssl3_get_record:bad record type:ssl/record/ssl3_record.c:371:
---
no peer certificate available
---
No client certificate CA names sent
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 138 bytes and written 304 bytes
Verification: OK
---
New, TLSv1.3, Cipher is TLS_CHACHA20_POLY1305_SHA256
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
```
This patch ensures that all modified ciphersuites must be from the FIPS provider (if enabled) and any unsupported ciphersuites are rejected.

Expected result:
```
$ openssl s_server -accept 127.0.0.1:4433 -cert cert.pem -provider fips -propquery fips=yes -ciphersuites TLS_CHACHA20_POLY1305_SHA256
Using default temp DH parameters
ACCEPT
ERROR
40A7902A967F0000:error:0A0000B5:SSL routines:tls_setup_handshake:no ciphers available:ssl/statem/statem_lib.c:170:No ciphers enabled for max supported SSL/TLS version
shutting down SSL
CONNECTION CLOSED
```
```
$ LD_LIBRARY_PATH=/opt/openssl-3.1.0/lib64 /opt/openssl-3.1.0/bin/openssl s_client -connect 127.0.0.1:4433 
Connecting to 127.0.0.1
CONNECTED(00000003)
40E752849F7F0000:error:0A000410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl/record/rec_layer_s3.c:1584:SSL alert number 40
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 297 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
```
CLA: regular

### CODE REVIEW

**Review by t8m — APPROVED** [review]

**Review by kaduk — APPROVED** [review]
I guess one could claim that this is fixing an omission from #12037...

File: ssl/ssl_ciph.c [hunk 1]
```diff
@@ -1390,9 +1391,17 @@ static int update_cipher_list(STACK_OF(SSL_CIPHER) **cipher_list,
...
-                             sk_SSL_CIPHER_value(tls13_ciphersuites, i), i);
+    for (i = sk_SSL_CIPHER_num(tls13_ciphersuites) - 1; i >= 0; i--) {
+        const SSL_CIPHER *sslc = sk_SSL_CIPHER_value(tls13_ciphersuites, i);
+
+        /* Don't include any TLSv1.3 ciphersuites that are disabled */
+        if ((sslc->algorithm_enc & ctx->disabled_enc_mask) == 0
+                && (ssl_cipher_table_mac[sslc->algorithm2
+                                         & SSL_HANDSHAKE_MAC_MASK].mask
+                    & ctx->disabled_mac_mask) == 0) {
+            sk_SSL_CIPHER_unshift(tmp_cipher_list, sslc);
```
Comment (paulidale) [inline ssl/ssl_ciph.c:1402]: Why the change from `sk_SSL_CIPHER_insert` to `sk_SSL_CIPHER_unshift`?

Comment (kaduk) [reply ssl/ssl_ciph.c:1402]: The index is no longer guaranteed to match up between the input and ouput stacks.

Comment (paulidale) [reply ssl/ssl_ciph.c:1402]: Wouldn't `sk_SSL_CIPHER_push` make more sense though?  The unshift function involves a lot of memory movement, push doesn't.

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Merged to both branches, thanks for the contribution.
